### PR TITLE
Fix: Improve text encoding for shell output in VSCode preview (#6178)

### PR DESCRIPTION
**Pull request by LIHUA919**
## 🐛 Problem

Users running commands with non-ASCII characters (like Russian text "пример") in Windows/WSL environments experience garbled text in VSCode's shell preview window, with Unicode replacement characters (�) appearing instead of the actual text.

**Issue**: https://github.com/openai/codex/issues/6178

## 🔧 Root Cause

The issue was in `StreamOutput<Vec<u8>>::from_utf8_lossy()` method in `codex-rs/core/src/exec.rs`, which used `String::from_utf8_lossy()` to convert shell output bytes to strings. This function immediately replaces any invalid UTF-8 byte sequences with replacement characters, without attempting to decode using other common encodings.

In Windows/WSL environments, shell output often uses encodings like:

- Windows-1252 (common Windows encoding)
- Latin-1/ISO-8859-1 (extended ASCII)

## 🛠️ Solution

Replaced the simple `String::from_utf8_lossy()` call with intelligent encoding detection via a new `bytes_to_string_smart()` function that tries multiple encoding strategies:

1. **UTF-8** (fast path for valid UTF-8)
2. **Windows-1252** (handles Windows-specific characters in 0x80-0x9F range)
3. **Latin-1** (fallback for extended ASCII)
4. **Lossy UTF-8** (final fallback, same as before)

### New Files

- `codex-rs/core/src/text_encoding.rs` - Smart encoding detection module
- `codex-rs/core/tests/suite/text_encoding_fix.rs` - Integration tests

### Modified Files

- `codex-rs/core/src/lib.rs` - Added text_encoding module
- `codex-rs/core/src/exec.rs` - Updated StreamOutput::from_utf8_lossy()
- `codex-rs/core/tests/suite/mod.rs` - Registered new test module

## ✅ Testing

- **5 unit tests** covering UTF-8, Windows-1252, Latin-1, and fallback scenarios
- **2 integration tests** simulating the exact Issue #6178 scenario
- **Demonstrates improvement** over the previous `String::from_utf8_lossy()` approach

All tests pass:

```bash
cargo test -p codex-core text_encoding
cargo test -p codex-core test_shell_output_encoding_issue_6178
```

## 🎯 Impact

- ✅ **Eliminates garbled text** in VSCode shell preview for non-ASCII content
- ✅ **Supports Windows/WSL environments** with proper encoding detection
- ✅ **Zero performance impact** for UTF-8 text (fast path)
- ✅ **Backward compatible** - UTF-8 content works exactly as before
- ✅ **Handles edge cases** with robust fallback mechanism

## 🧪 Test Scenarios

The fix has been tested with:

- Russian text ("пример")
- Windows-1252 quotation marks (""test")
- Latin-1 accented characters ("café")
- Mixed encoding content
- Invalid byte sequences (graceful fallback)

## 📋 Checklist

- [X] Addresses the reported issue
- [X] Includes comprehensive tests
- [X] Maintains backward compatibility
- [X] Follows project coding conventions
- [X] No breaking changes

### CODE REVIEW

**Comment by etraut-openai** [comment]
@codex review

**Comment by etraut-openai** [comment]
Thanks for the contribution!

Please sign the CLA, fix the CI failures (looks like a simple formatting issue), and take a look at the codex code review feedback.

**Comment by LIHUA919** [comment]
I have read the CLA Document and I hereby sign the CLA

File: codex-rs/core/src/text_encoding.rs [hunk 1]
```diff
@@ -0,0 +1,151 @@
...
+            0x00..=0x7F => byte as char, // ASCII range
+            0x80..=0x9F => WINDOWS_1252_MAP[(byte - 0x80) as usize], // Windows-1252 specific
+            0xA0..=0xFF => byte as char, // ISO-8859-1 compatible range
+        };
+        result.push(ch);
+    }
+
+    // Validate that the result makes sense (contains reasonable characters)
+    if result.chars().any(|c| c.is_control() && c != '\n' && c != '\r' && c != '\t') {
+        return None;
```
[the remark this answers is not in the document]
Comment (joshka-oai) [reply codex-rs/core/src/text_encoding.rs]: This is still better than what was there before. The problem mentioned by codex is correct. This PR narrows the current failure mode to 1252/latin + ANSI sequences. Ideally shell commands that are run non-interactively should omit formatting when possible, so it's likely this is a much smaller portion of the whole.

That said, a fairly trivial fix for this might be to add \x1b to the allowed character lists here.

File: codex-rs/core/src/text_encoding.rs
Comment (joshka-oai) [inline codex-rs/core/src/text_encoding.rs:1]: Is there an existing rust library for this that we can use rather than owning this code?

File: codex-rs/core/src/text_encoding.rs (hunk 1 above)
[the remark this answers is not in the document]
Comment (joshka-oai) [reply codex-rs/core/src/text_encoding.rs]: (Looks like this was updated with that fix in mind already)

**Review by joshka-oai — COMMENTED** [review]
General approval with the approach and code changes from a visual inspection. I haven't yet tested this however, but will later this week.

File: codex-rs/core/src/text_encoding.rs
Comment (joshka-oai) [reply codex-rs/core/src/text_encoding.rs:1]: Maybe encoding-rs?

**Comment by joshka-oai** [comment]
@codex review

I've modified this to use encoding_rs and chardetng (another crate by the same developer as encoding_rs that detects the character sets). Codex rewrote most of the unit tests and added some explicit handling for codepage 866 / Windows 1252 oddness - I tried to get it to find a canonical link to some page that documents the problem it was solving here, but it could only really say that there was a problem and point at the two codepages as evidence. The logic seems reasonable here though. I'm not an expert however on those things.

@LIHUA919 @nos1609 would you mind taking a quick look at this. I can see that unit tests are reasonably detailed and seemingly cover the previous concerns, but this is one of those places where the real world tends to have some edge cases that are difficult to anticipate.

**Comment by joshka-oai** [comment]
@codex review

File: codex-rs/core/src/text_encoding.rs [hunk 2]
```diff
@@ -0,0 +1,158 @@
...
+                return false;
+            }
+        }
+        if byte.is_ascii_alphabetic() {
+            saw_ascii_word = true;
+        }
+    }
+
+    saw_extended_punctuation && saw_ascii_word
+}
```
[the remark this answers is not in the document]
Comment (joshka-oai) [reply codex-rs/core/src/text_encoding.rs:113]: https://github.com/openai/codex/pull/6182/commits/e4bc1627b9f5f935cb6568a658b7952e85578d5e

File: codex-rs/core/src/text_encoding.rs [hunk 3]
```diff
@@ -0,0 +1,182 @@
...
+            return false;
+        }
+        if (0x80..=0x9F).contains(&byte) {
+            if !WINDOWS_1252_PUNCT_BYTES.contains(&byte) {
+                return false;
+            }
+            saw_extended_punctuation = true;
+            high_byte_count += 1;
+            if high_byte_count > 4 {
+                return false;
```
[the remark this answers is not in the document]
Comment (joshka-oai) [reply codex-rs/core/src/text_encoding.rs]: [9a48531](https://github.com/openai/codex/pull/6182/commits/9a485311925bd47c87fc0abd0fa68f99187b16ee)

**Comment by joshka-oai** [comment]
@codex review

File: codex-rs/core/src/text_encoding.rs [hunk 4]
```diff
@@ -0,0 +1,199 @@
...
+// to preserve. Windows-1252 byte values for smart punctuation.
+const WINDOWS_1252_PUNCT_BYTES: [u8; 8] = [
+    0x91, // ‘ (left single quotation mark)
+    0x92, // ’ (right single quotation mark)
+    0x93, // “ (left double quotation mark)
+    0x94, // ” (right double quotation mark)
+    0x95, // • (bullet)
+    0x96, // – (en dash)
+    0x97, // — (em dash)
+    0x99, // ™ (trade mark sign)
```
[the remark this answers is not in the document]
Comment (joshka-oai) [reply codex-rs/core/src/text_encoding.rs:46]: I'm going to push back against this one for now - we could chase all the various characters here. Let's wait for an actual problem to occur to re-review which characters should be in this list.

**Review by joshka-oai — APPROVED** [review]

**Comment by nos1609** [comment]
@joshka-oai The only doubt actually from my point of view is the lack of Windows 1251 codepage. The more kind-of "universal" would be to handle the Linux's [LatArCyrHeb](https://developer.ibm.com/articles/l-u-cyr/) approch to cover non Latin-charaters output, especially cyrillic ones.

The Windows 1252 looks like not covering the ArCyrHeb segments of the caracter maps.

Just a small "hello world" example
The word "Привет" in Windows 1252 would be rendered as "ÐŸÑ€Ð¸Ð²ÐµÑ‚" which is not really human-legible as "Hello" because of lack of cyrillic characters from Windows 1251.

https://learn.microsoft.com/windows/win32/intl/code-page-identifiers

PS if got it right, this covers the case
https://github.com/openai/codex/pull/6182/files#diff-6603ce56d5786a455399c6dd3791b120a4e22c65ae19e0f924d1f8a4901cd60cR132-R144

**Comment by joshka-oai** [comment]
@nos1609 disregard the main PR comment here - I got codex to reimplement this from the manual detection into a more automated one using https://crates.io/crates/chardetng and https://crates.io/crates/encoding_rs. These should handle detecting 1251 and a large variety of other codepages. I'll add your test here as a new unit test to make sure that scenario is covered correctly though.

---

To test this in your VS Code extension:

Rebuild codex locally (cargo build) and edit your settings.json file to point to it. Something like this (adjust the path as appropriate):

```
"chatgpt.cliExecutable": "/Users/<user name>/code/codex/codex-rs/target/debug/codex",
```

Then when you restart the extension, it will use this CLI executable. You'll see a little orange icon in the lower right corner of the extension panel to remind you that you're using a custom CLI. Make sure to remove this when you're done testing to ensure that you get up to date versions of the CLI when updating the extention.
